### PR TITLE
Task/graphs v1

### PR DESCRIPTION
**Pull request by alonisser**

### CODE REVIEW

File: uncontrast_studies/processors/parameters_distribution_bar.py [hunk 1]
```diff
@@ -0,0 +1,183 @@
...
+        experiments_subquery_by_breakdown = self.filtered_experiments.filter(paradigm__main=OuterRef("pk")).values(
+            "id"
+        )  # TODO: adapt this one according to the final modeling
+
+        breakdown_query = UnConMainParadigm.objects.values("name").distinct().annotate(series_name=F("name"))
+
+        qs = self._aggregate_query_by_breakdown(breakdown_query, experiments_subquery_by_breakdown)
+        return qs
+
+    def process_population(self):
```
Comment (dani-efr) [inline uncontrast_studies/processors/parameters_distribution_bar.py:43]: I prefer naming it "process_sample_type" to be more consistent with the model name

File: uncontrast_studies/processors/parameters_distribution_bar.py [hunk 2]
```diff
@@ -0,0 +1,183 @@
...
+        ).values("id")
+
+        breakdown_query = UnConStimulusCategory.objects.values("name").distinct().annotate(series_name=F("name"))
+
+        qs = self._aggregate_query_by_breakdown(breakdown_query, experiments_subquery_by_breakdown)
+        return qs
+
+    def process_suppressed_stimuli_modality(self):
+        experiments_subquery_by_breakdown = self.filtered_experiments.filter(
+            suppressed__stimuli__modality=OuterRef("pk")
```
Comment (dani-efr) [inline uncontrast_studies/processors/parameters_distribution_bar.py:97]: shouldn't it be `suppressed_stimuli__modality` ?

File: uncontrast_studies/processors/parameters_distribution_bar.py [hunk 3]
```diff
@@ -0,0 +1,183 @@
...
+        ).values("id")
+
+        breakdown_query = UnConModalityType.objects.values("name").distinct().annotate(series_name=F("name"))
+
+        qs = self._aggregate_query_by_breakdown(breakdown_query, experiments_subquery_by_breakdown)
+        return qs
+
+    def process_target_stimuli_modality(self):
+        experiments_subquery_by_breakdown = self.filtered_experiments.filter(
+            suppressed__stimuli__modality=OuterRef("pk")
```
Comment (dani-efr) [inline uncontrast_studies/processors/parameters_distribution_bar.py:107]: same question

File: uncontrast_studies/processors/parameters_distribution_bar.py [hunk 4]
```diff
@@ -0,0 +1,183 @@
...
+        ).values("experiment")
+
+        breakdown_query = UnConProcessingMainDomain.objects.values("name").distinct().annotate(series_name=F("name"))
+
+        qs = self._aggregate_query_by_breakdown(breakdown_query, experiments_subquery_by_breakdown)
+        return qs
+
+    def process_processing_domain(self):
+        experiments_subquery_by_breakdown = self.filtered_experiments.filter(
+            experiment__processing_domains__main=OuterRef("pk")
```
Comment (dani-efr) [inline uncontrast_studies/processors/parameters_distribution_bar.py:147]: why start with experiment here, not like all the other processors?

File: uncontrast_studies/processors/parameters_distribution_free_queries.py [hunk 5]
```diff
@@ -0,0 +1,239 @@
...
+        self.consciousness_measure_phases = kwargs.pop("consciousness_measure_phases", [])
+        self.consciousness_measure_types = kwargs.pop("consciousness_measure_types", [])
+        self.processing_domain_main_types = kwargs.pop("processing_domain_types", [])
+        self.suppression_methods_types = kwargs.pop("suppression_methods_types", [])
+        self.types = kwargs.pop("types", [])
+        self.tasks = kwargs.pop("tasks", [])
+
+    def process(self):
+        """
+        do a transpose "experiment per country in study" with unnest on the array field
```
Comment (dani-efr) [inline uncontrast_studies/processors/parameters_distribution_free_queries.py:43]: this also seems unrelated

File: uncontrast_studies/processors/parameters_distribution_free_queries.py [hunk 6]
```diff
@@ -0,0 +1,239 @@
...
+
+    def process_paradigm(self):
+        experiments_subquery_by_breakdown = self.filtered_experiments.filter(paradigm__main=OuterRef("pk")).values("id")
+
+        breakdown_query = UnConMainParadigm.objects.values("name").distinct().annotate(series_name=F("name"))
+
+        qs = self._aggregate_query_by_breakdown(breakdown_query, experiments_subquery_by_breakdown)
+        return qs
+
+    def process_population(self):
```
Comment (dani-efr) [inline uncontrast_studies/processors/parameters_distribution_free_queries.py:104]: same naming problem as before

File: uncontrast_studies/processors/parameters_distribution_free_queries.py [hunk 7]
```diff
@@ -0,0 +1,239 @@
...
+        ).values("id")
+
+        breakdown_query = UnConStimulusCategory.objects.values("name").distinct().annotate(series_name=F("name"))
+
+        qs = self._aggregate_query_by_breakdown(breakdown_query, experiments_subquery_by_breakdown)
+        return qs
+
+    def process_suppressed_stimuli_modality(self):
+        experiments_subquery_by_breakdown = self.filtered_experiments.filter(
+            suppressed__stimuli__modality=OuterRef("pk")
```
Comment (dani-efr) [inline uncontrast_studies/processors/parameters_distribution_free_queries.py:158]: same __ problem

File: uncontrast_studies/processors/parameters_distribution_free_queries.py [hunk 8]
```diff
@@ -0,0 +1,239 @@
...
+        ).values("id")
+
+        breakdown_query = UnConModalityType.objects.values("name").distinct().annotate(series_name=F("name"))
+
+        qs = self._aggregate_query_by_breakdown(breakdown_query, experiments_subquery_by_breakdown)
+        return qs
+
+    def process_target_stimuli_modality(self):
+        experiments_subquery_by_breakdown = self.filtered_experiments.filter(
+            suppressed__stimuli__modality=OuterRef("pk")
```
Comment (dani-efr) [inline uncontrast_studies/processors/parameters_distribution_free_queries.py:168]: same

File: uncontrast_studies/processors/parameters_distribution_free_queries.py [hunk 9]
```diff
@@ -0,0 +1,239 @@
...
+        ).values("experiment")
+
+        breakdown_query = UnConProcessingMainDomain.objects.values("name").distinct().annotate(series_name=F("name"))
+
+        qs = self._aggregate_query_by_breakdown(breakdown_query, experiments_subquery_by_breakdown)
+        return qs
+
+    def process_processing_domain(self):
+        experiments_subquery_by_breakdown = self.filtered_experiments.filter(
+            experiment__processing_domains__main=OuterRef("pk")
```
Comment (dani-efr) [inline uncontrast_studies/processors/parameters_distribution_free_queries.py:208]: same question about experiment

File: uncontrast_studies/processors/parameters_distribution_pie.py [hunk 10]
```diff
@@ -0,0 +1,263 @@
+import itertools
+
+from django.contrib.postgres.expressions import ArraySubquery
+from django.db.models import QuerySet, OuterRef, F, Func, Count
+from django.db.models.functions import JSONObject
+
+from contrast_api.choices import InterpretationsChoices
+from studies.models import (
```
Comment (dani-efr) [inline uncontrast_studies/processors/parameters_distribution_pie.py:8]: why do you import all the contrast models here?

Comment (dani-efr) [reply uncontrast_studies/processors/parameters_distribution_pie.py:8]: also all following methods seem to relate to previous models

File: uncontrast_studies/processors/trends_over_time.py [hunk 11]
```diff
@@ -0,0 +1,185 @@
...
+
+    def process_paradigm(self):
+        experiments_subquery_by_breakdown = self.experiments.filter(paradigm__main=OuterRef("pk"))
+
+        breakdown_query = UnConMainParadigm.objects.values("name").distinct("name").annotate(series_name=F("name"))
+
+        qs = self._aggregate_query_by_breakdown(breakdown_query, experiments_subquery_by_breakdown)
+        return qs
+
+    def process_population(self):
```
Comment (dani-efr) [inline uncontrast_studies/processors/trends_over_time.py:42]: same naming issue

File: uncontrast_studies/processors/trends_over_time.py [hunk 12]
```diff
@@ -0,0 +1,185 @@
...
+    def process_task(self):
+        experiments_subquery_by_breakdown = self.experiments.filter(tasks__type=OuterRef("pk"))
+
+        breakdown_query = UnConTaskType.objects.values("name").distinct("name").annotate(series_name=F("name"))
+
+        qs = self._aggregate_query_by_breakdown(breakdown_query, experiments_subquery_by_breakdown)
+        return qs
+
+    def process_processing_domain(self):
+        experiments_subquery_by_breakdown = self.experiments.filter(experiment__processing_domains__main=OuterRef("pk"))
```
Comment (dani-efr) [inline uncontrast_studies/processors/trends_over_time.py:73]: same

File: uncontrast_studies/serializers.py [hunk 13]
```diff
@@ -0,0 +1,210 @@
...
+        fields = ("experiment", "id", "type")
+
+
+class UnConsciousnessMeasureSerializer(serializers.ModelSerializer):
+    phase = serializers.PrimaryKeyRelatedField(queryset=UnConsciousnessMeasurePhase.objects.all())
+    type = serializers.PrimaryKeyRelatedField(queryset=UnConsciousnessMeasureType.objects.all())
+
+    class Meta:
+        model = UnConsciousnessMeasure
+        fields = ("experiment", "id", "phase", "type")
```
Comment (dani-efr) [inline uncontrast_studies/serializers.py:49]: what about sub-type? and metadata?

File: uncontrast_studies/views/experiments_graphs.py [hunk 14]
```diff
@@ -0,0 +1,223 @@
...
+        "parameters_distribution_bar": ParametersDistributionBarGraphDataProcessor,
+        "parameters_distribution_pie": ParametersDistributionPieGraphDataProcessor,
+        "parameters_distribution_free_queries": ParametersDistributionFreeQueriesDataProcessor,
+    }
+
+    @extend_schema(
+        responses=NationOfConsciousnessGraphSerializer(many=True),
+        parameters=[
+            OpenApiParameter(
+                name="theory", type=str, required=False, many=True, description="theory filter - supports multiple"
```
Comment (dani-efr) [inline uncontrast_studies/views/experiments_graphs.py:87]: again with the theory related graphs...

File: uncontrast_studies/processors/base.py [hunk 15]
```diff
@@ -0,0 +1,37 @@
...
+from uncontrast_studies.models import UnConExperiment
+
+
+class BaseProcessor:
+    def __init__(self, experiments: QuerySet[UnConExperiment], **kwargs):
+        self.experiments: QuerySet[UnConExperiment] = experiments
+        min_number_of_experiments = int(kwargs.pop("min_number_of_experiments", ["0"])[0])
+        if min_number_of_experiments > 0:
+            # so everything that explicitly sets a minimum it would be supported, with a gt query
+            min_number_of_experiments = min_number_of_experiments - 1
```
[the remark this answers is not in the document]
Comment (alonisser) [reply uncontrast_studies/processors/base.py:15]: כתוב למעלה. אבל זה מהמקורי.. 
            # so everything that explicitly sets a minimum it would be supported, with a gt query

File: uncontrast_studies/processors/journals.py [hunk 16]
```diff
@@ -0,0 +1,44 @@
+from django.db.models import QuerySet, Count, F
+
+from uncontrast_studies.models import UnConExperiment
+from uncontrast_studies.processors.base import BaseProcessor
+
+
+class JournalsGraphDataProcessor(BaseProcessor):
+    """
+    This expects as input the ID of the parent theory..
```
[the remark this answers is not in the document]
Comment (alonisser) [reply uncontrast_studies/processors/journals.py:9]: אכן אני אעיף את ההערה

File: uncontrast_studies/processors/nations_of_consciousness.py [hunk 17]
```diff
@@ -0,0 +1,59 @@
...
+
+
+class NationOfConsciousnessDataProcessor(BaseProcessor):
+    def __init__(self, experiments: QuerySet[UnConExperiment], **kwargs):
+        super().__init__(experiments=experiments, **kwargs)
+
+    def process(self):
+        """
+        do a transpose "experiment per country in study" with unnest on the array field
+        aggregate on country and theory relation
```
[the remark this answers is not in the document]
Comment (alonisser) [reply uncontrast_studies/processors/nations_of_consciousness.py:14]: yap, I'll add a todo to change this

File: uncontrast_studies/processors/nations_of_consciousness.py [hunk 18]
```diff
@@ -0,0 +1,59 @@
...
+        experiments_by_countries = self.get_queryset()
+
+        aggregate = self.aggregate(experiments_by_countries)
+
+        return aggregate
+
+    def get_queryset(self):
+        filtered_qs = self.experiments
+
+        experiments_by_countries_and_theories = (
```
[the remark this answers is not in the document]
Comment (alonisser) [reply uncontrast_studies/processors/nations_of_consciousness.py:28]: true, but it would change when we get the depdendt  right

File: uncontrast_studies/processors/parameters_distribution_bar.py (hunk 1 above)
Comment (alonisser) [reply uncontrast_studies/processors/parameters_distribution_bar.py:43]: I know, but it's from the "original" system, and in this point we're stuck with it

File: uncontrast_studies/processors/trends_over_time.py (hunk 12 above)
Comment (alonisser) [reply uncontrast_studies/processors/trends_over_time.py:73]: same

File: uncontrast_studies/serializers.py (hunk 13 above)
Comment (alonisser) [reply uncontrast_studies/serializers.py:49]: true, I'll add it

File: uncontrast_studies/views/experiments_graphs.py (hunk 14 above)
Comment (alonisser) [reply uncontrast_studies/views/experiments_graphs.py:87]: right I'll fix it
